trace_decoder/src/protocol_processing.rs

```diff
@@ -0,0 +1,173 @@
+//! Logic to processes the "outer" structure of the protocol payload. All
+//! "inner" complex logic (eg. compact decoding) is handled in separate
+//! dedicated modules.
+
+use std::collections::HashMap;
+
+use enum_as_inner::EnumAsInner;
+use ethereum_types::{Address, U256};
+use mpt_trie::partial_trie::HashedPartialTrie;
+use thiserror::Error;
+
+use crate::{
+    compact::compact_prestate_processing::{
+        process_compact_prestate_debug, MptPartialTriePreImages,
+    },
+    decoding_mpt::{MptTraceParsingError, MptTraceParsingResult},
+    decoding_smt::{SmtTraceParsingError, SmtTraceParsingResult},
+    processed_block_trace::ProcessedBlockTrace,
+    processed_block_trace_mpt::{MptProcessedBlockTracePreImages, ProcessingMeta},
+    processed_block_trace_smt::SmtProcessedBlockTracePreImages,
+    trace_protocol::{
+        BlockTrace, MptBlockTraceTriePreImages, MptCombinedPreImages,
+        MptSeparateStorageTriesPreImage, MptSeparateTriePreImage, MptSeparateTriePreImages,
+        MptTrieCompact, MptTrieUncompressed, SingleSmtPreImage, SmtBlockTraceTriePreImages,
+        TriePreImage,
+    },
+    types::HashedAccountAddr,
+};
+
+const COMPATIBLE_HEADER_VERSION: u8 = 1;
+
+pub type TraceParsingResult<T> = Result<T, TraceParsingError>;
+
+// TODO: Split into two error types based on if `UnexpectedTraceFormat` can
```

Comment:
What do you mean by can occur?

Reply:
So maybe I should remove this, but I added a function that works with both `mpt` & `smt` inputs (`process_block_trace_trie_pre_images`). However, most of the time I expect that the caller will only expect to process `mpt` or `smt` payloads.

`UnexpectedTraceFormat` can only occur when the caller uses `process_mpt_block_trace_trie_pre_images` & `process_smt_block_trace_trie_pre_images`. They all return the same error type though, so `process_block_trace_trie_pre_images` still has the `UnexpectedTraceFormat` even though it can never create it.